kirkwood: add support for Netgear Stora (MS2110) NAS

wigyori opened the pull request:
Dual-slot NAS based on Marvell Kirkwood.

Specifications:
 - Marvell 88F6281 @1GHz
 - 128Mb RAM
 - 256Mb NAND
 - 1x GbE LAN (Marvell 88E1116)
 - 1x USB 2.0
 - 2x SATA
 - PCF8563 RTC
 - LM75 sensor
 - TC654 PWM fan controller
 - Serial on J2 (115200,8n1)
 - Newer bootROM so kwboot-ing via serial is possible

Installation:

1. Serial console
 - Connect your levelshifter to the serial console
   on J2 (refer to the wiki page for pinout)
2. Update u-boot
 - Download the u-boot.kwb image for the device
 - Powercycle the NAS
 - Run "kwboot -b ./u-boot.kwb /dev/ttyUSB0 -p"
 - Connect to the serial console with minicom
 - tftp 0x0800000 netgear_stora-u-boot.kwb
 - nand erase 0x0 100000
 - nand write 0x0800000 0x0 0x100000
 - reset
3. Install OpenWrt
 - Boot up the initramfs image
 - tftpboot 0x800000 openwrt-kirkwood-netgear_stora-initramfs-uImage; bootm 0x800000
 - Download the sysupgrade image and perform sysupgrade

The fan is controlled in 3 stages by a script running every minute
from cron, measuring the CPU temperature.

Snippets taken from bodhi <mibodhi@gmail.com>

Signed-off-by: Zoltan HERPAI <wigyori@uid0.hu>

adschm commented:
> modules: add hwmon-tc654

This should be "kirkwood: add hwmon-tc654 as module" or similar

adschm commented on target/linux/kirkwood/modules.mk:
```diff
@@ -0,0 +1,20 @@
+#
+# Copyright (C) 2006-2014 OpenWrt.org
```
This should be updated to 2020.

Personally, I'd remove it from such a trivial file, though.

adschm commented on target/linux/kirkwood/files-5.4/arch/arm/boot/dts/kirkwood-stora.dts:
```diff
@@ -0,0 +1,220 @@
+// SPDX-License-Identifier: GPL-2.0+
```
GPL-2.0-or-later

adschm commented on target/linux/kirkwood/files-5.4/arch/arm/boot/dts/kirkwood-stora.dts:
```diff
@@ -0,0 +1,220 @@
+
+/ {
+	model = "NETGEAR MS2000 / MS2110 aka Stora";
+	compatible = "netgear,stora", "netgear,ms2000", "netgear,ms2110", "marvell,kirkwood-88f6281", "marvell,kirkwood";
```
I'd prefer the specific name here, something like

```
	model = "NETGEAR Stora MS2110";
	compatible = "netgear,stora-ms2110", "marvell,kirkwood-88f6281", "marvell,kirkwood";
```

or

```
	model = "NETGEAR MS2110";
	compatible = "netgear,ms2110", "marvell,kirkwood-88f6281", "marvell,kirkwood";
```

Netgear seems to use the precise name as well.

wigyori replied:
Netgear used MS2000 for the packages not including a disk, and MS2110 for the packages including a disk, so it's hard to make a distinction. As more sources refer to this as MS2110, I'll pick that one.

RaylynnKnight replied:
A better option for maintaining consistency with existing kirkwood dts files would be to rename this file to kirkwood-netgear_stora.dts  See the existing dts files for the ReadyNAS Duo (kirkwood-netgear_duo_v2.dts)  and ReadyNAS NV+ (kirkwood-netgear_nv+_v2.dts).

adschm commented on target/linux/kirkwood/files-5.4/arch/arm/boot/dts/kirkwood-stora.dts:
```diff
@@ -0,0 +1,220 @@
+		pinctrl-names = "default";
+
+		power_led {
+			label = "status:blue:power_led";
```
Those labels do not match the color:function scheme. Anything preventing us from just dropping the "status:" prefix?

adschm replied:
And, apart from that, we could also remove the "-led" suffix, ending up with
blue:power
green:disk1
etc.

as we do on other devices as well ...

adschm commented on target/linux/kirkwood/files-5.4/arch/arm/boot/dts/kirkwood-stora.dts:
```diff
@@ -0,0 +1,220 @@
+		pinctrl-0 = <&pmx_button_power &pmx_button_reset >;
+		pinctrl-names = "default";
+
+		power-button {
```
Just "power" will be enough, this is a member of the keys node so we know it's a "button".

adschm commented on target/linux/kirkwood/files-5.4/arch/arm/boot/dts/kirkwood-stora.dts:
```diff
@@ -0,0 +1,220 @@
+};
+
+&pciec {
+        status = "okay";
```
spaces->tabs for indent.

adschm commented on target/linux/kirkwood/image/Makefile:
```diff
@@ -136,6 +136,15 @@ define Device/linksys_ea4500
 
+define Device/netgear_stora
+  DEVICE_VENDOR := Netgear
+  DEVICE_MODEL := Stora
```
See above, I think either "Stora MS2110" or "MS2110 (Stora)" will better match the name as used by then vendor.

CHKDSK88 commented on target/linux/kirkwood/files-5.4/arch/arm/boot/dts/kirkwood-stora.dts:
```diff
@@ -0,0 +1,220 @@
+			      &pmx_led_red_disk1 &pmx_led_red_disk2 >;
+		pinctrl-names = "default";
+
+		power_led {
```
No aliases with status leds?

mans0n commented on target/linux/kirkwood/patches-5.4/112-netgear_stora.patch:
```diff
@@ -0,0 +1,10 @@
+ 	kirkwood-net5big.dtb \
+ 	kirkwood-netgear_readynas_duo_v2.dtb \
+ 	kirkwood-netgear_readynas_nv+_v2.dtb \
++	kirkwood-stora.dtb \
```
We don't need this patch, it will be compiled automatically.

mans0n commented on target/linux/kirkwood/image/Makefile:
```diff
@@ -136,6 +136,15 @@ define Device/linksys_ea4500
 TARGET_DEVICES += linksys_ea4500
 
+define Device/netgear_stora
+  DEVICE_VENDOR := Netgear
```
Here "NETGEAR" in all caps is prefered.

mans0n commented on target/linux/kirkwood/files-5.4/arch/arm/boot/dts/kirkwood-stora.dts:
```diff
@@ -0,0 +1,227 @@
+	status = "okay";
+
+	ethphy0: ethernet-phy@0 { /* Marvell 88E1318 */
+		compatible = "marvell,88e1116";
```
Probably you don't need this line. https://github.com/torvalds/linux/blob/master/drivers/net/mdio/of_mdio.c#L181

wigyori replied:
Good point, thanks.

mans0n commented on target/linux/kirkwood/files-5.4/arch/arm/boot/dts/kirkwood-stora.dts:
```diff
@@ -0,0 +1,227 @@
+&mdio {
+	status = "okay";
+
+	ethphy0: ethernet-phy@0 { /* Marvell 88E1318 */
```
Shouldn't we use `ethernet-phy@8` here?

mans0n commented on target/linux/kirkwood/files-5.4/arch/arm/boot/dts/kirkwood-stora.dts:
```diff
@@ -0,0 +1,227 @@
+
+	partition@e0000 {
+		label = "u-boot-env";
+		reg = <0x0e00000 0x020000>;
```
`0x0e0000`?
It would be nicer if you use same number of digits (e.g. 8) or remove leading zeros entirely.

mans0n commented on target/linux/kirkwood/base-files/etc/uci-defaults/04_storafan:
```diff
@@ -0,0 +1,22 @@
+board=$(board_name)
+
+case "$board" in
+netgear,stora)
```
If it can be extended to accommodate other devices as well, then why use the board-specific filename?

wigyori replied:
> If it can be extended to accommodate other devices as well, then why use the board-specific filename?

mvebu has a similar script (mvebu/cortexa9/base-files/etc/uci-defaults/04_mambafan). Some unification might be done with these, but I'm not in a position to change mvebu.

chunkeey replied:
Note: The Netgear WNDR4700 uses a tc654 too. However instead of a script, it uses thermal-zones in the device-tree to control the fan. (With a special boot script /etc/init.d/hwmon_fancontrol to put the chip into pwm mode)

mans0n commented:
The hwmon-tc654 module is not a target-specific driver, IMO it would be better to add it to `package/kernel/linux/modules/hwmon.mk` directly.

RaylynnKnight commented on target/linux/kirkwood/files-5.4/arch/arm/boot/dts/kirkwood-stora.dts:
```diff
@@ -0,0 +1,227 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
```
Can this file be renamed for consistency with existing kirkwood-netgear files?  The Netgear ReadyNAS Duo v2 (kirkwood-netgear_readynas_duo_v2.dts) and Ready NAS NV+ v2 (kirkwood-netgear_readynas_nv+_v2.dts) would indicate that kirkwood-netgear_stora.dts would be a more consistent name.

wigyori replied:
The kernel already has the kirkwood-linksys-viper.dtsi while we have the kirkwood-e4200-v2 / kirkwood-ea3500 / kirkwood-ea4500 DTS files. I don't see this as an odd-ball here for the missing vendor name, but don't mind changing it.

adschm replied:
Since kernel does not care about consistent names at all, I'd also not follow that example. _If_ netgear should be in the name, we should probably go for kirkwood-netgear-stora.dts.

wigyori replied:
> The hwmon-tc654 module is not a target-specific driver, IMO it would be better to add it to `package/kernel/linux/modules/hwmon.mk` directly.

I've broken it out to the standard hwmon.mk then, and then the apm821xx target might want to consider using it in its nand profile.

wigyori commented:
Ping @adschm / @mans0n

apcameron commented:
@adschm @mans0n any comments/updates

wigyori commented:
Ping.

wigyori commented:
Rebased on master.

ozerkov-k commented on target/linux/kirkwood/files-5.4/arch/arm/boot/dts/kirkwood-stora.dts:
```diff
@@ -0,0 +1,226 @@
+
+		led_power: power {
+			label = "blue:power";
+			gpios = <&gpio0 31 GPIO_ACTIVE_LOW>;
```
What is correlation with pmx_led_blue_power (mpp32) in line 123 ?

AFAIK Stora has two GPIOs for a blue led:
mpp31 <-> gpio0 31 -- full bright
mpp32 <-> gpio1 0 -- half bright

wigyori replied:
Good point, thanks, I've changed the pmx definition to use the full bright LED.

apcameron commented:
@wigyori any updates on this PR to resolve the branch conflicts

wigyori commented:
> @wigyori any updates on this PR to resolve the branch conflicts

Will rebase when 5.15 is in for kirkwood.

wigyori committed:
kirkwood: add support for Netgear Stora (MS2000/2110) NAS

Dual-slot NAS based on Marvell Kirkwood.

Specifications:
 - Marvell 88F6281 @1GHz
 - 128Mb RAM
 - 256Mb NAND
 - 1x GbE LAN (Marvell 88E1116)
 - 1x USB 2.0
 - 2x SATA
 - PCF8563 RTC
 - LM75 sensor
 - TC654 PWM fan controller
 - Serial on J2 (115200,8n1)
 - Newer bootROM so kwboot-ing via serial is possible

Installation:

1. Serial console
 - Connect your levelshifter to the serial console
   on J2 (refer to the wiki page for pinout)
2. Update u-boot
 - Download the u-boot.kwb image for the device
 - Powercycle the NAS
 - Run "kwboot -b ./u-boot.kwb /dev/ttyUSB0 -p"
 - Connect to the serial console with minicom
 - tftp 0x0800000 netgear_stora-u-boot.kwb
 - nand erase 0x0 100000
 - nand write 0x0800000 0x0 0x100000
 - reset
3. Install OpenWrt
 - Boot up the initramfs image
 - tftpboot 0x800000 openwrt-kirkwood-netgear_stora-initramfs-uImage; bootm 0x800000
 - Download the sysupgrade image and perform sysupgrade

The fan is controlled in 3 stages by a script running every minute
from cron, measuring the CPU temperature.

Snippets taken from bodhi <mibodhi@gmail.com>

Signed-off-by: Zoltan HERPAI <wigyori@uid0.hu>

wigyori commented:
> > @wigyori any updates on this PR to resolve the branch conflicts
> 
> Will rebase when 5.15 is in for kirkwood.

PR rebased.

apcameron commented:
@wigyori It looks like the checks have failed. Please can you review what's wrong and address it.

CHKDSK88 commented on target/linux/kirkwood/modules.mk:
```diff
@@ -0,0 +1,20 @@
+# See /LICENSE for more information.
+#
+
+define KernelPackage/hwmon-tc654
```
Why it's target specific package? It could be used in different devices too.

wigyori replied:
There is a clash with the apm821xx target (ping @chunkeey for opinions). It has SENSORS_TC654 compiled into the kernel (and the tc654 driver has patches added in apm821xx/patches-5.15/010-hwmon-tc654-add-thermal_cooling-device.patch) . 

I agree that the target-specific hwmon-tc654 could/should be moved to be a generic one, but then the apm821xx target would also need to be updated to use the module, or I also have to compile in the tc654 driver - which means that the driver would be enabled for the whole kirkwood target as there are no subtargets there. Given that most if not all kirkwood boards are NAND-based, the size increase would likely be close to be taken as a rounding error, so I'm okay with either.

apcameron replied:
@wigyori @adschm  It would be great to have this PR accepted soon as the Legacy Stora software has not been receiving any security updates for a few years now.

chunkeey replied:
clash? Do you have a build failure or something? 

From what I know, the kmod-hwmon-tc654 will just not contain a .ko for the apm821xx. But this there's otherwise no real-life consequences. 

The NETGEAR WNDR4700 SoC is running hotter because it has a 200MHz higher clock than any other APM821xx device. The thermal management is handled by the thermal-zone framework itself, that's why it's "neat" to have the fan-controller+temperature sensor built-in.

RaylynnKnight commented:
What is the status on this pull request?  Can we get it merged?  If it still needs work can someone summarize what is needed?

apcameron commented:
@wigyori Any updates on this PR?

RaylynnKnight reviewed (approved):
3 years since original submission.  Can someone either point out what are the current issues with the pull request and/or finally merge this please?!!

wigyori committed:
hwmon: add TC654 fan-controller support

Add support for the Microchip TC654 PWM fan-controller.

Signed-off-by: Zoltan HERPAI <wigyori@uid0.hu>

wigyori commented:
Merged, thanks all for the comments and reviews.
